### PR TITLE
implement pgd for tfv2 and pytorch

### LINKED ISSUE #344: Investigate performance of Projected Gradient Descent (PGD)
Investigate and compare time and accuracy of Projected Gradient Descent (PGD) implementations.

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
Implement PGD attacks for specific frameworks such as TFv2 and Pytorch.

Fixes #344 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/__init__.py [hunk 1]
```diff
@@ -3,29 +3,28 @@
...
-# from art.attacks.evasion.pixel_threshold import PixelAttack
-# from art.attacks.evasion.pixel_threshold import ThresholdAttack
-# from art.attacks.evasion.frame_saliency import FrameSaliencyAttack
-#
-# from art.attacks.poisoning.backdoor_attack import PoisoningAttackBackdoor
-# from art.attacks.poisoning.poisoning_attack_svm import PoisoningAttackSVM
-#
-# from art.attacks.extraction.functionally_equivalent_extraction import FunctionallyEquivalentExtraction
-# from art.attacks.extraction.copycat_cnn import CopycatCNN
+from art.attacks.evasion.adversarial_patch.adversarial_patch import AdversarialPatch
```
Comment (beat-buesser) [inline art/attacks/__init__.py:6]: Please keep these imports commented.

File: art/attacks/evasion/projected_gradient_descent/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from art.attacks.evasion.projected_gradient_descent.projected_gradient_descent import ProjectedGradientDescent
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/__init__.py:1]: I think we should leave this file empty and use internally the full import paths instead.

File: art/attacks/__init__.py [hunk 3]
```diff
@@ -3,29 +3,28 @@
...
+from art.attacks.evasion.virtual_adversarial import VirtualAdversarialMethod
+from art.attacks.evasion.zoo import ZooAttack
+from art.attacks.evasion.pixel_threshold import PixelAttack
+from art.attacks.evasion.pixel_threshold import ThresholdAttack
+
+from art.attacks.poisoning.backdoor_attack import PoisoningAttackBackdoor
+from art.attacks.poisoning.poisoning_attack_svm import PoisoningAttackSVM
+
+from art.attacks.extraction.functionally_equivalent_extraction import FunctionallyEquivalentExtraction
+from art.attacks.extraction.copycat_cnn import CopycatCNN
```
Comment (beat-buesser) [inline art/attacks/__init__.py:30]: Please keep these imports commented.
```suggestion
# from art.attacks.evasion.adversarial_patch.adversarial_patch import AdversarialPatch
# from art.attacks.evasion.boundary import BoundaryAttack
# from art.attacks.evasion.carlini import CarliniL2Method, CarliniLInfMethod
# from art.attacks.evasion.decision_tree_attack import DecisionTreeAttack
# from art.attacks.evasion.deepfool import DeepFool
# from art.attacks.evasion.elastic_net import ElasticNet
# from art.attacks.evasion.fast_gradient import FastGradientMethod
# from art.attacks.evasion.hclu import HighConfidenceLowUncertainty
# from art.attacks.evasion.hop_skip_jump import HopSkipJump
# from art.attacks.evasion.iterative_method import BasicIterativeMethod
# from art.attacks.evasion.newtonfool import NewtonFool
# from art.attacks.evasion.projected_gradient_descent.projected_gradient_descent import ProjectedGradientDescent
# from art.attacks.evasion.saliency_map import SaliencyMapMethod
# from art.attacks.evasion.spatial_transformation import SpatialTransformation
# from art.attacks.evasion.universal_perturbation import UniversalPerturbation
# from art.attacks.evasion.virtual_adversarial import VirtualAdversarialMethod
# from art.attacks.evasion.zoo import ZooAttack
# from art.attacks.evasion.pixel_threshold import PixelAttack
# from art.attacks.evasion.pixel_threshold import ThresholdAttack

# from art.attacks.poisoning.backdoor_attack import PoisoningAttackBackdoor
# from art.attacks.poisoning.poisoning_attack_svm import PoisoningAttackSVM

# from art.attacks.extraction.functionally_equivalent_extraction import FunctionallyEquivalentExtraction
# from art.attacks.extraction.copycat_cnn import CopycatCNN
```

File: art/attacks/evasion/projected_gradient_descent/__init__.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+from art.attacks.evasion.projected_gradient_descent.projected_gradient_descent_numpy import \
+    ProjectedGradientDescentNumpy
+from art.attacks.evasion.projected_gradient_descent.projected_gradient_descent_pytorch import \
+    ProjectedGradientDescentPytorch
+from art.attacks.evasion.projected_gradient_descent.projected_gradient_descent_tensorflow_v2 import \
+    ProjectedGradientDescentTensorflowV2
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/__init__.py:7]: I think we should leave this file empty and use internally the full import paths instead.
```suggestion
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            "max_iter": max_iter,
+            "targeted": targeted,
+            "num_random_init": num_random_init,
+            "batch_size": batch_size,
+            "random_eps": random_eps
+        }
+        ProjectedGradientDescent.set_params(self, **kwargs)
+
+        if isinstance(self.estimator, PyTorchClassifier):
+            self._attack = ProjectedGradientDescentPytorch(
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:123]:
```suggestion
            self._attack = ProjectedGradientDescentPyTorch(
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+                eps_step=eps_step,
+                max_iter=max_iter,
+                targeted=targeted,
+                num_random_init=num_random_init,
+                batch_size=batch_size,
+                random_eps=random_eps
+            )
+
+        elif isinstance(self.estimator, TensorFlowV2Classifier):
+            self._attack = ProjectedGradientDescentTensorflowV2(
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:136]:
```suggestion
            self._attack = ProjectedGradientDescentTensorFlowV2(
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 7]
```diff
@@ -136,6 +135,7 @@ def generate(self, x, y=None, **kwargs):
             self.eps = np.round(self.norm_dist.rvs(1)[0], 10)
             self.eps_step = ratio * self.eps
 
+        mask = kwargs.get("mask")
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:138]: Why did you move this line here? If it is to have it at the beginning, why not put it right after the docstring?

File: tests/attacks/test_projected_gradient_descent.py [hunk 8]
```diff
@@ -134,14 +137,8 @@ def _test_backend_mnist(self, classifier, x_train, y_train, x_test, y_test):
...
         self.assertAlmostEqual(float(np.max(np.abs(x_test_original - x_test))), 0.0, delta=0.00001)
 
-        # Test the masking:
-        attack = ProjectedGradientDescent(classifier, num_random_init=1)
-        mask = np.random.binomial(n=1, p=0.5, size=np.prod(x_test.shape))
-        mask = mask.reshape(x_test.shape)
-
-        x_test_adv = attack.generate(x_test, mask=mask)
-        mask_diff = (1 - mask) * (x_test_adv - x_test)
-        self.assertAlmostEqual(float(np.max(np.abs(mask_diff))), 0.0, delta=0.00001)
```
Comment (beat-buesser) [inline tests/attacks/test_projected_gradient_descent.py:144]: Why are the tests for masking removed?

File: tests/attacks/test_projected_gradient_descent.py [hunk 9]
```diff
@@ -294,8 +291,240 @@ def test_scikitlearn(self):
...
+    def test_framework_tensorflow_v2_mnist(self):
+        classifier, _ = get_image_classifier_tf()
+
+        scores = get_labels_np_array(classifier.predict(self.x_train_mnist))
+        acc = np.sum(np.argmax(scores, axis=1) == np.argmax(self.y_train_mnist, axis=1)) / self.y_train_mnist.shape[0]
+        logger.info("[TFV2, MNIST] Accuracy on training set: %.2f%%", acc * 100)
+
+        scores = get_labels_np_array(classifier.predict(self.x_test_mnist))
+        acc = np.sum(np.argmax(scores, axis=1) == np.argmax(self.y_test_mnist, axis=1)) / self.y_test_mnist.shape[0]
+        logger.info("[TFV2, MNIST] Accuracy on test set: %.2f%%", acc * 100)
```
Comment (beat-buesser) [inline tests/attacks/test_projected_gradient_descent.py:304]: I think we can remove these tests, because they are not testing PGD.

File: tests/attacks/test_projected_gradient_descent.py [hunk 10]
```diff
@@ -294,8 +291,240 @@ def test_scikitlearn(self):
...
+        self.x_test_mnist = np.swapaxes(self.x_test_mnist, 1, 3).astype(np.float32)
+        classifier = get_image_classifier_pt()
+
+        scores = get_labels_np_array(classifier.predict(self.x_train_mnist))
+        acc = np.sum(np.argmax(scores, axis=1) == np.argmax(self.y_train_mnist, axis=1)) / self.y_train_mnist.shape[0]
+        logger.info("[PyTorch, MNIST] Accuracy on training set: %.2f%%", acc * 100)
+
+        scores = get_labels_np_array(classifier.predict(self.x_test_mnist))
+        acc = np.sum(np.argmax(scores, axis=1) == np.argmax(self.y_test_mnist, axis=1)) / self.y_test_mnist.shape[0]
+        logger.info("[PyTorch, MNIST] Accuracy on test set: %.2f%%", acc * 100)
```
Comment (beat-buesser) [inline tests/attacks/test_projected_gradient_descent.py:319]: I think we can remove these tests, because they are not testing PGD.

File: tests/attacks/test_projected_gradient_descent.py [hunk 11]
```diff
@@ -294,8 +291,240 @@ def test_scikitlearn(self):
...
+            num_random_init=0,
+            batch_size=2,
+            random_eps=False,
+        )
+        x_train_adv_fw = attack_fw.generate(self.x_train_mnist)
+        x_test_adv_fw = attack_fw.generate(self.x_test_mnist)
+
+        # Test
+        self.assertAlmostEqual(np.mean(x_train_adv_np), np.mean(x_train_adv_fw), places=6)
+        self.assertAlmostEqual(np.mean(x_test_adv_np), np.mean(x_test_adv_fw), places=6)
```
Comment (beat-buesser) [inline tests/attacks/test_projected_gradient_descent.py:358]: How about testing the difference, that way we can make sure that data has been modified.
```suggestion
        self.assertAlmostEqual(np.mean(x_train_adv_np - self.x_train_mnist), np.mean(x_train_adv_fw - self.x_train_mnist), places=6)
        self.assertAlmostEqual(np.mean(x_test_adv_np - self.x_test_mnist), np.mean(x_test_adv_fw - self.x_test_mnist), places=6)
```

File: tests/attacks/test_projected_gradient_descent.py [hunk 12]
```diff
@@ -294,8 +291,240 @@ def test_scikitlearn(self):
...
```
Comment (beat-buesser) [inline tests/attacks/test_projected_gradient_descent.py:391]:
```suggestion
        self.assertAlmostEqual(np.mean(x_train_adv_np - self.x_train_mnist), np.mean(x_train_adv_fw - self.x_train_mnist), places=6)
        self.assertAlmostEqual(np.mean(x_test_adv_np - self.x_test_mnist), np.mean(x_test_adv_fw - self.x_test_mnist), places=6)
```

File: art/estimators/classification/tensorflow.py [hunk 13]
```diff
@@ -910,6 +910,34 @@ def class_gradient(self, x, label=None, **kwargs):
...
+        :type x: `Tensor`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,).
+        :type y: `np.ndarray`
+        :return: Gradients of the same shape as `x`.
+        :rtype: `Tensor`
+        """
+        import tensorflow as tf
+
+        if tf.executing_eagerly():
```
Comment (beat-buesser) [inline art/estimators/classification/tensorflow.py:927]: What happens if TensorFlow is not executing eagerly? Should we raise an exception like NotImplementedError?

File: art/estimators/classification/tensorflow.py [hunk 14]
```diff
@@ -910,6 +910,34 @@ def class_gradient(self, x, label=None, **kwargs):
...
+        :param x: Input with shape as expected by the model.
+        :type x: `Tensor`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,).
+        :type y: `np.ndarray`
+        :return: Gradients of the same shape as `x`.
+        :rtype: `Tensor`
+        """
+        import tensorflow as tf
+
```
Comment (beat-buesser) [inline art/estimators/classification/tensorflow.py:926]: I think in the framework-specific implementations we are currently not applying preprocessing and preprocessor defences. I think we should add checks somewhere to warn the user, for example if the provided classifier has preprocessing and preprocessor defences defined.

File: art/estimators/classification/pytorch.py [hunk 15]
```diff
@@ -395,6 +404,58 @@ def loss_gradient(self, x, y, **kwargs):
...
+
+        # Check label shape
+        if self._reduce_labels:
+            y = torch.argmax(y, dim=1)
+
+        # Convert the inputs to Variable
+        x = Variable(x, requires_grad=True)
+
+        # Compute the gradient and return
+        model_outputs = self._model(x)
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:430]: Should we call self._model.train() or self._model.eval() to make sure the model is in the right mode for gradient calculation?

File: art/estimators/classification/pytorch.py [hunk 16]
```diff
@@ -395,6 +404,58 @@ def loss_gradient(self, x, y, **kwargs):
...
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Input with shape as expected by the model.
+        :type x: `Tensor`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,).
+        :type y: `Tensor`
+        :return: Gradients of the same shape as `x`.
+        :rtype: `Tensor`
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:417]: Is this correct?
```suggestion
        :rtype: `torch.Tensor`
```

File: art/estimators/classification/pytorch.py [hunk 17]
```diff
@@ -395,6 +404,58 @@ def loss_gradient(self, x, y, **kwargs):
 
         return grads
 
+    def loss_gradient_framework(self, x, y, **kwargs):
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Input with shape as expected by the model.
+        :type x: `Tensor`
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:412]: Is this correct?
```suggestion
        :type x: `torch.Tensor`
```

File: art/estimators/classification/pytorch.py [hunk 18]
```diff
@@ -123,6 +123,15 @@ def __init__(
             self._reduce_labels = False
 
     @property
+    def get_device(self):
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:126]: If we make this a property we can just call it `device`.
```suggestion
    def device(self):
```

File: art/estimators/classification/tensorflow.py [hunk 19]
```diff
@@ -910,6 +910,34 @@ def class_gradient(self, x, label=None, **kwargs):
...
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Input with shape as expected by the model.
+        :type x: `Tensor`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,).
+        :type y: `np.ndarray`
+        :return: Gradients of the same shape as `x`.
+        :rtype: `Tensor`
```
Comment (beat-buesser) [inline art/estimators/classification/tensorflow.py:923]:
```suggestion
        :rtype: `tensorflow.Tensor`
```

File: art/estimators/classification/tensorflow.py [hunk 20]
```diff
@@ -910,6 +910,34 @@ def class_gradient(self, x, label=None, **kwargs):
...
 
+    def loss_gradient_framework(self, x, y, **kwargs):
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Input with shape as expected by the model.
+        :type x: `Tensor`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,).
+        :type y: `np.ndarray`
```
Comment (beat-buesser) [inline art/estimators/classification/tensorflow.py:921]:
```suggestion
        :type y: `tensorflow.Tensor`
```

File: art/estimators/classification/tensorflow.py [hunk 21]
```diff
@@ -910,6 +910,34 @@ def class_gradient(self, x, label=None, **kwargs):
 
         return gradients
 
+    def loss_gradient_framework(self, x, y, **kwargs):
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Input with shape as expected by the model.
+        :type x: `Tensor`
```
Comment (beat-buesser) [inline art/estimators/classification/tensorflow.py:918]:
```suggestion
        :type x: `tensorflow.Tensor`
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py [hunk 22]
```diff
@@ -0,0 +1,450 @@
...
+            raise NotImplementedError(
+                "Values of `norm_p` different from 1, 2 and `np.inf` are currently not supported."
+            )
+
+        values = values_tmp.reshape(values.shape)
+
+        return values
+
+    def set_params(self, **kwargs):
+        """
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:400]: It could be that the `set_params` methods of all ProjectedGradientDescent* are the same or very similar. Would it be possible to define it in`ProjectedGradientDescent` and reuse it in `ProjectedGradientDescentNumpy`, `ProjectedGradientDescentTensorFlow` and `ProjectedGradientDescentPyTorch`?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py [hunk 23]
```diff
@@ -0,0 +1,450 @@
...
+
+from art.config import ART_NUMPY_DTYPE
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.attacks.attack import EvasionAttack
+from art.utils import compute_success, get_labels_np_array, check_and_transform_label_format, random_sphere
+
+logger = logging.getLogger(__name__)
+
+
+class ProjectedGradientDescentPytorch(EvasionAttack):
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:42]: Would it be possible to inherit from `ProjectedGradientDescent` and only overwrite the method `generate`? That way we could reuse for example code in the constructor.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py [hunk 24]
```diff
@@ -0,0 +1,450 @@
...
+        rate_best = None
+
+        for _ in range(max(1, self.num_random_init)):
+            adv_x = x.astype(ART_NUMPY_DTYPE)
+
+            # Compute perturbation with implicit batching
+            for batch_id in range(int(np.ceil(x.shape[0] / float(self.batch_size)))):
+                batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+                batch = x[batch_index_1:batch_index_2]
+                batch_labels = targets[batch_index_1:batch_index_2]
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:174]: Would it be now possible to use `torch.utils.data` for the batching? It could improve performance.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 25]
```diff
@@ -0,0 +1,447 @@
...
+        rate_best = None
+
+        for _ in range(max(1, self.num_random_init)):
+            adv_x = x.astype(ART_NUMPY_DTYPE)
+
+            # Compute perturbation with implicit batching
+            for batch_id in range(int(np.ceil(x.shape[0] / float(self.batch_size)))):
+                batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+                batch = x[batch_index_1:batch_index_2]
+                batch_labels = targets[batch_index_1:batch_index_2]
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:174]: Would it be now possible to use `tf.data.Dataset` for the batching? It could improve performance.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 26]
```diff
@@ -0,0 +1,447 @@
...
+        else:
+            raise NotImplementedError(
+                "Values of `norm_p` different from 1, 2 and `np.inf` are currently not supported."
+            )
+
+        values = tf.reshape(values_tmp, values.shape)
+
+        return values
+
+    def set_params(self, **kwargs):
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:396]: See comment about set_params for `projected_gradient_descent_pytorch.py`.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 27]
```diff
@@ -0,0 +1,447 @@
...
+
+from art.config import ART_NUMPY_DTYPE
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.attacks.attack import EvasionAttack
+from art.utils import compute_success, get_labels_np_array, check_and_transform_label_format, random_sphere
+
+logger = logging.getLogger(__name__)
+
+
+class ProjectedGradientDescentTensorflowV2(EvasionAttack):
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:42]: See comment about parent class for `projected_gradient_descent_pytorch.py`.

Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:42]:
```suggestion
class ProjectedGradientDescentTensorFlowV2(EvasionAttack):
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 23 above)
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:42]:
```suggestion
class ProjectedGradientDescentPyTorch(EvasionAttack):
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py (hunk 7 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:138]: In the numpy version, this line is put within the `if` clause and is referenced in the `else` clause, so should be corrected by putting it before the `if` clause, could be right after the docstring.

File: tests/attacks/test_projected_gradient_descent.py (hunk 8 above)
Comment (minhitbk) [reply tests/attacks/test_projected_gradient_descent.py:144]: This is accidentially removed, it is now recovered.

File: tests/attacks/test_projected_gradient_descent.py (hunk 9 above)
Comment (minhitbk) [reply tests/attacks/test_projected_gradient_descent.py:304]: Done.

File: tests/attacks/test_projected_gradient_descent.py (hunk 10 above)
Comment (minhitbk) [reply tests/attacks/test_projected_gradient_descent.py:319]: Done.

File: art/estimators/classification/tensorflow.py (hunk 13 above)
Comment (minhitbk) [reply art/estimators/classification/tensorflow.py:927]: Done.

File: art/estimators/classification/tensorflow.py (hunk 14 above)
Comment (minhitbk) [reply art/estimators/classification/tensorflow.py:926]: Added the warning in the init func of each framework specific implementation.

File: art/estimators/classification/pytorch.py (hunk 15 above)
Comment (minhitbk) [reply art/estimators/classification/pytorch.py:430]: No, we already do that in the set_learning_phase(self, train) of the classifier.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 22 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:400]: No, we cannot do that because there are no inheritance relationship between the main PGD and different implementations of it.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 23 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py]: No, unless we change the current design of the use of self._attack.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 24 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:174]: No, we need numpy as input for those framework-specific attack.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 25 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:174]: No, we need numpy as input for the attack.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 26 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:396]: No, there are no inheritance relationship among the attacks.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 27 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py]: No, see explain above.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 22 above)
Comment (mathsinn) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:411]: Might it be worthwhile then to create a new parent class for `ProjectedGradientDescentNumpy`, `ProjectedGradientDescentTensorFlow` and `ProjectedGradientDescentPyTorch` to have the common logic (in particular from `__init__` and `set_params`, as well as the checks at the beginning of `generate`) in one place?

Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:411]: I have created a new parent class to include this comment.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 23 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py]: Yes, this is done with a new parent class for all implementation of PGD.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 26 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:407]: Yes, this is done with a new parent class for all implementation of PGD.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 27 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py]: Yes, this is done with a new parent class for all implementation of PGD.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 25 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:148]: Done.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 24 above)
Comment (minhitbk) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:148]: Done.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py [hunk 28]
```diff
@@ -0,0 +1,235 @@
...
+    def __init__(
+        self,
+        estimator,
+        norm=np.inf,
+        eps=0.3,
+        eps_step=0.1,
+        max_iter=100,
+        targeted=False,
+        num_random_init=0,
+        batch_size=1,
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:82]:
```suggestion
        batch_size=32,
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 29]
```diff
@@ -103,13 +104,178 @@ def __init__(
...
+    def __init__(
+        self,
+        estimator,
+        norm=np.inf,
+        eps=0.3,
+        eps_step=0.1,
+        max_iter=100,
+        targeted=False,
+        num_random_init=0,
+        batch_size=1,
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:238]:
```suggestion
        batch_size=32,
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py [hunk 30]
```diff
@@ -0,0 +1,392 @@
...
+    def __init__(
+        self,
+        estimator,
+        norm=np.inf,
+        eps=0.3,
+        eps_step=0.1,
+        max_iter=100,
+        targeted=False,
+        num_random_init=0,
+        batch_size=1,
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:61]:
```suggestion
        batch_size=32,
```

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 31]
```diff
@@ -0,0 +1,381 @@
...
+    def __init__(
+        self,
+        estimator,
+        norm=np.inf,
+        eps=0.3,
+        eps_step=0.1,
+        max_iter=100,
+        targeted=False,
+        num_random_init=0,
+        batch_size=1,
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:61]:
```suggestion
        batch_size=32,
```

**Comment by beat-buesser** [comment]
Thank you @minhitbk and @mathsinn !
